Add OPN instruction and vararg operator table

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27104 Add OPN instruction and vararg operator table**

* The use case here is to replace prim::ListConstruct, which requires Node, but Node is not available in mobile lite interpreter. 
* (OPN, X, N), X is the index to the vararg operator-name and operator tables. N is number of inputs. For ListConstruct example, operator name can be "aten::listconstruct" and the overloaded name is the output type ("int", "float", "bool", "tensor" and "generic").
* A vararg operator table is built with void(int input_size, Stack& stack) functions. 
## Unit test
LiteInterpreterConv covers OPN instruction and conv operator.

Differential Revision: [D17762853](https://our.internmc.facebook.com/intern/diff/D17762853)

Commit:
Add OPN instruction. Register operators for computer vision models.

Commit:
Update on "Add OPN instruction. Register operators for computer vision models."


## Add OPN instruction. 
* The use case here is to replace prim::ListConstruct, which requires Node. Node is not available in mobile lite interpreter. 
* (OPN, X, N), X is the index to the operator-name and operator tables. N is number of inputs. For ListConstruct example, operator name can be "aten::listconstruct" and the overloaded name is the output type ("int", "float", "bool", "tensor" and "generic").
## Register operators for cv models
Currently the operators need to be registered manually through c10 op registration, in register_mobile_ops.cpp.
## Tested models
* Lenet
* xrayMobile v3
## Unit test
LiteInterpreterConv covers OPN instruction and conv operator.

zdevito commented on torch/csrc/jit/instruction.h:
```diff
@@ -19,6 +19,7 @@ namespace jit {
 
 #define FORALL_OPCODES(_)                                                   \
   _(OP, "O") /* invoke operator X */                                        \
+  _(OPN, "OI") /* invoke operator X and an interger value N */              \
```
/* invoke vararg operator X with N arguments

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -400,6 +400,10 @@ struct CodeImpl {
     return instructions_;
   }
 
+  const std::vector<Node*>& instructions_source() const {
```
If we have instructions_source then we probably do not need opname_table since the operator name can be extracted from the Node.

zdevito commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -1,25 +1,42 @@
+
+  //
   auto opname = code_->op_names_.back();
+  // Special treatments for some prim ops.
```
This function should probably just not be called for OPN instructions. A separate append_vararg_operator should be called that does the resolution into a separate vararg table.

zdevito commented on torch/csrc/jit/mobile/interpreter.cpp:
```diff
@@ -28,13 +34,35 @@ bool InterpreterState::run(Stack& stack) {
         c10::Dispatcher::singleton().callBoxed(*code_->operators_[inst.X], &stack);
         ++pc;
       } break;
+      case OPN: {
```
This isn't the right place to do this resolution. There should be a code_->vararg_operators_ table that holds `void fn(int N, Stack stack)` functions that this instruction invokes.

zdevito commented on torch/csrc/jit/mobile/register_mobile_ops.cpp:
```diff
@@ -23,5 +23,138 @@ static auto registry0 = torch::RegisterOperators().op(
   [](at::Tensor a, at::Scalar b, at::Scalar c) ->at::Tensor {
     return at::add(a, b, c);
   })
+).op(
```
These changes seem like a different PR

iseeyuan replied:
They will be moved to a new PR. This PR would only include the operators that are needed in unit test.

zdevito reviewed (commented):
Some inline comments, general idea is in the right direction.

Commit:
Update on "Add OPN instruction. Register operators for computer vision models."


## Add OPN instruction. 
* The use case here is to replace prim::ListConstruct, which requires Node. Node is not available in mobile lite interpreter. 
* (OPN, X, N), X is the index to the operator-name and operator tables. N is number of inputs. For ListConstruct example, operator name can be "aten::listconstruct" and the overloaded name is the output type ("int", "float", "bool", "tensor" and "generic").
## Register operators for cv models
Currently the operators need to be registered manually through c10 op registration, in register_mobile_ops.cpp.
## Tested models
* Lenet
* xrayMobile v3
## Unit test
LiteInterpreterConv covers OPN instruction and conv operator.

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -668,12 +668,41 @@ class ScriptModuleSerializer {
+            } else if (lt->getElementType() == BoolType::get()) {
+              opnames[ins.X].overload_name = "bool";
+            } else if (lt->getElementType()->isSubtypeOf(TensorType::get())) {
+              opnames[ins.X].overload_name = "tensor";
```
nit: Tensor, to match the actual type as it appears in Python

zdevito reviewed (approved):
Looks good! Thanks for making the changes.

Commit:
Update on "Add OPN instruction and vararg operator table"


* The use case here is to replace prim::ListConstruct, which requires Node, but Node is not available in mobile lite interpreter. 
* (OPN, X, N), X is the index to the vararg operator-name and operator tables. N is number of inputs. For ListConstruct example, operator name can be "aten::listconstruct" and the overloaded name is the output type ("int", "float", "bool", "tensor" and "generic").
* A vararg operator table is built with void(int input_size, Stack& stack) functions. 
## Unit test
LiteInterpreterConv covers OPN instruction and conv operator.

Commit:
Update on "Add OPN instruction and vararg operator table"


* The use case here is to replace prim::ListConstruct, which requires Node, but Node is not available in mobile lite interpreter. 
* (OPN, X, N), X is the index to the vararg operator-name and operator tables. N is number of inputs. For ListConstruct example, operator name can be "aten::listconstruct" and the overloaded name is the output type ("int", "float", "bool", "tensor" and "generic").
* A vararg operator table is built with void(int input_size, Stack& stack) functions. 
## Unit test
LiteInterpreterConv covers OPN instruction and conv operator.

pietern commented:
@iseeyuan There are clang-tidy problems with this PR.

iseeyuan commented:
@pietern The tidy problems happen in register_mobile_ops.cpp, which is a short-term solution for op registration in mobile. It will be replaced by auto registration in near future, when ATen to c10 migration is complete. I'll try to clean them up for next PR.
